src/bn256.cpp

```diff
@@ -1,411 +1,331 @@
-      temp.marshal(ret[6]);
+   affined.x_.x_.mont_decode().marshal(view.subspan<0, num_bytes>());
+   affined.x_.y_.mont_decode().marshal(view.subspan<num_bytes, num_bytes>());
+   affined.y_.x_.mont_decode().marshal(view.subspan<num_bytes * 2, num_bytes>());
+   affined.y_.y_.mont_decode().marshal(view.subspan<num_bytes * 3, num_bytes>());
+}
 
-      temp.mont_decode(p_.y_.x_.y_);
-      temp.marshal(ret[7]);
+// unmarshal sets g2 to the result of converting the output of marshal back into
+// a group element and then returns unmarshal_status.
+std::error_code g2::unmarshal(std::span<const uint8_t, 128> m) noexcept {
+   constexpr auto num_bytes = 256 / 8;
+
+   if (auto ec = p().x_.x_.unmarshal(m.subspan<0, num_bytes>()); ec != unmarshal_error::NO_ERROR)
+      return ec;
+
+   if (auto ec = p().x_.y_.unmarshal(m.subspan<num_bytes, num_bytes>()); ec != unmarshal_error::NO_ERROR)
+      return ec;
+
+   if (auto ec = p().y_.x_.unmarshal(m.subspan<num_bytes * 2, num_bytes>()); ec != unmarshal_error::NO_ERROR)
+      return ec;
+
+   if (auto ec = p().y_.y_.unmarshal(m.subspan<num_bytes * 3, num_bytes>()); ec != unmarshal_error::NO_ERROR)
+      return ec;
+
+   p().x_.x_ = p().x_.x_.mont_encode();
+   p().x_.y_ = p().x_.y_.mont_encode();
+   p().y_.x_ = p().y_.x_.mont_encode();
+   p().y_.y_ = p().y_.y_.mont_encode();
+
+   if (p().x_.is_zero() && p().y_.is_zero()) {
+      // This is the point at infinity.
+      p().y_.set_one();
+      p().z_.set_zero();
+      p().t_.set_zero();
+   } else {
+      p().z_.set_one();
+      p().t_.set_one();
+      if (!p().is_on_curve()) {
+         return unmarshal_error::MALFORMED_POINT;
+      }
+   }
 
-      temp.mont_decode(p_.y_.y_.x_);
-      temp.marshal(ret[8]);
+   return {};
+}
 
-      temp.mont_decode(p_.y_.y_.y_);
-      temp.marshal(ret[9]);
+std::string gt::string() const { return p().string(); }
 
-      temp.mont_decode(p_.y_.z_.x_);
-      temp.marshal(ret[10]);
 
-      temp.mont_decode(p_.y_.z_.y_);
-      temp.marshal(ret[11]);
+gt::gt(const gfp12& p) {
+   static_assert(sizeof(*this) == sizeof(gfp12));
+   memcpy(this, &p, sizeof(*this));
+}
 
-      return ret;
-   }
+gfp12& gt::p() {
+   static_assert(sizeof(*this) == sizeof(gfp12));
+   return *reinterpret_cast<gfp12*>(p_);
+}
 
-   unmarshal_status gt::unmarshal(uint8_array_32_array_12_t& m) {
-      unmarshal_status status = p_.x_.x_.x_.unmarshal(m[0]);
-      if (status != unmarshal_success) {
-         return status;
-      }
+const gfp12& gt::p() const {
+   static_assert(sizeof(*this) == sizeof(gfp12));
+   return *reinterpret_cast<const gfp12*>(p_);
+}
 
-      status = p_.x_.x_.y_.unmarshal(m[1]);
-      if (status != unmarshal_success) {
-         return status;
-      }
+// scalar_mult return a*k
+gt gt::scalar_mult(const int512_t& k) const noexcept { return gt{ p().exp(k) }; }
+
+// add sets gt to a+b and then returns gt.
+gt gt::add(const gt& b) const noexcept { return gt{ p().mul(b.p()) }; }
```

Comment:
why `p().mul` not `p().add`?

Reply:
Typo, will change it.

Reply:
It was fallowing the golang code

Reply:
// Add sets e to a+b and then returns e.
func (e *GT) Add(a, b *GT) *GT {
	if e.p == nil {
		e.p = &gfP12{}
	}
	e.p.Mul(a.p, b.p)
	return e
}

Reply:
We could rename the method to mul and update it's use in the main_test.cpp TEST_CASE("test_pairings", "[main]")

Reply:
Actually, this is not typo. This is corresponding to https://github.com/ethereum/go-ethereum/blob/fb75f11e87420ec25ff72f7eeeb741fa8974e87e/crypto/bn256/cloudflare/bn256.go#L360